engine/src/main/java/org/hibernate/search/engine/backend/work/execution/OperationSubmitter.java

```diff
@@ -8,52 +8,128 @@
 
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.RejectedExecutionException;
+import java.util.function.Consumer;
+import java.util.function.Function;
 
+import org.hibernate.search.engine.common.execution.spi.SimpleScheduledExecutor;
 import org.hibernate.search.util.common.annotation.Incubating;
 
 
 /**
- * Interface defining how operation should be submitted to the queue.
+ * Interface defining how operation should be submitted to the queue or executor.
+ * <p>
+ * <strong>WARNING:</strong> while this type is API, because instances should be manipulated by users,
+ * all of its methods are considered SPIs and therefore should never be called directly by users.
+ * In short, users are only expected to get instances of this type from an API and pass it to another API.
+ * <p>
+ * Currently supported implementations:
+ * <ul>
+ *     <li>{@link #blocking()}</li>
+ *     <li>{@link #rejectedExecutionException()}</li>
+ *     <li>{@link #offloading(Consumer)}</li>
+ * </ul>
  */
 @Incubating
-public enum OperationSubmitter {
+public abstract class OperationSubmitter {
+
+	private static final OperationSubmitter BLOCKING = new BlockingOperationSubmitter();
+	private static final OperationSubmitter REJECTED_EXECUTION_EXCEPTION = new RejectedExecutionExceptionOperationSubmitter();
+
+	private OperationSubmitter() {
+	}
+
+	/**
+	 * Defines how an element will be submitted to the queue.
+	 * Depending on the implementation might throw {@link RejectedExecutionException} or offload the submit operation to a provided executor.
+	 */
+	public abstract <T> void submitToQueue(BlockingQueue<? super T> queue, T element,
+			Consumer<? super T> blockingRetryProducer) throws InterruptedException;
+
+	/**
+	 * Defines how an element will be submitted to the executor.
+	 * Depending on the implementation might throw {@link RejectedExecutionException} or offload the submit operation to an offload executor.
+	 */
+	public abstract <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
+			Function<? super T, Runnable> blockingRetryProducer) throws InterruptedException;
```

Comment:
Ah, if we switch to `Consumer<? super T> blockingRetryProducer` for the other method, maybe we should here, too? :)